### PR TITLE
buildah: create WORKDIR with USER permissions

### PR DESCRIPTION
**Pull request by giuseppe**
Closes: https://github.com/containers/buildah/issues/3620

needs https://github.com/openshift/imagebuilder/pull/219

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by nalind** [comment]
This _needs_ a conformance test.

File: run_linux.go
Comment (TomSweeneyRedHat) [inline run_linux.go:199]: This should move down 20 or 30 lines towards line 225 where it's first referenced.

**Comment by flouthoc** [comment]
I am not sure but I am thinking of this approach where we can add `uid, gid` to `type StageExecutor` and `builder` can update `StageExecutor's` `uid, gid` field on every `Run(` so `func (s *StageExecutor) EnsureContainerPath(` https://github.com/containers/buildah/blob/main/imagebuildah/stage_executor.go#L1572 itself can take care of assigning correct ownership to workdir and we might not need any changes in `imagebuilder`.

PS: approach in this above comment it just an idea I don't know if there can be more blockers ahead.

**Comment by giuseppe** [comment]
it turned out to be more complicated than I thought, but tests are green are now

**Comment by rhatdan** [comment]
LGTM, need to get imagebuilder merged first and get @nalind Approval.

**Comment by giuseppe** [comment]
I'll need to re-vendor with the correct repo once https://github.com/openshift/imagebuilder/pull/219 is merged

File: add.go
Comment (nalind) [inline add.go:683]: Pretty sure this needs to either supply a `UIDMap` and `GIDMap`, or handle ID mapping itself.

Comment (giuseppe) [reply add.go:685]: I've taken it from the existing code (where I cannot spot any `UIDMap` and `GIDMap` conversion): https://github.com/containers/buildah/pull/3840/files#diff-f2e4566c6b7e38384283187aba6d7fd91e5ba8da2ffd0f849277bb76bff27fb3L1573

I've added it now

Comment (nalind) [reply add.go:685]: It's entirely possible it wasn't ever actually creating the directory here.  If we don't have a test for that, we should add one.

**Comment by giuseppe** [comment]
added another test

**Comment by rhatdan** [comment]
/hold cancel
@nalind @flouthoc PTAL

**Comment by flouthoc** [comment]
Changes LGTM but we need imagebuilder PR first.

File: go.mod
Comment (nalind) [inline go.mod:30]: Ideally this would not be a pinned commit, but a tag.

**Comment by TomSweeneyRedHat** [comment]
@giuseppe I'll retest again, but the tests are still not happy.  Can you take a look please when you get in?

**Comment by TomSweeneyRedHat** [comment]
Looks like a real error in the copier code: 

```
] === RUN   TestGetPermissionErrorNoChroot
[+3007s] === RUN   TestGetPermissionErrorNoChroot/ignore=false
[+3007s] === RUN   TestGetPermissionErrorNoChroot/ignore=true
[+3007s]     copier_linux_test.go:149: 
[+3007s]         	Error Trace:	copier_linux_test.go:149
[+3007s]         	Error:      	Not equal: 
[+3007s]         	            	expected: 2
[+3007s]         	            	actual  : 4
[+3007s]         	Test:       	TestGetPermissionErrorNoChroot/ignore=true
[+3007s]         	Messages:   	expected two readable items, got 4
[+3007s] --- FAIL: TestGetPermissionErrorNoChroot (0.54s)
[+3007s]     --- PASS: TestGetPermissionErrorNoChroot/ignore=false (0.27s)
[+3007s]     --- FAIL: TestGetPermissionErrorNoChroot/ignore=true (0.27s)
[+3007s] === RUN   TestGetPermissionErrorChroot
[+3007s] === RUN   TestGetPermissionErrorChroot/ignore=false
[+3007s] === RUN   TestGetPermissionErrorChroot/ignore=true
[+3007s]     copier_linux_test.go:149: 
[+3007s]         	Error Trace:	copier_linux_test.go:149
[+3007s]         	Error:      	Not equal: 
[+3007s]         	            	expected: 2
[+3007s]         	            	actual  : 4
[+3007s]         	Test:       	TestGetPermissionErrorChroot/ignore=true
[+3007s]         	Messages:   	expected two readable items, got 4
[+3007s] --- FAIL: TestGetPermissionErrorChroot (0.54s)
[+3007s]     --- PASS: TestGetPermissionErrorChroot/ignore=false (0.26s)
[+3007s]     --- FAIL: TestGetPermissionErrorChroot/ignore=true (0.28s)
```

**Comment by TomSweeneyRedHat** [comment]
Also, if you have to adjust, can you vendor  imagebuilder to  v1.2.3 please?

File: go.mod
Comment (flouthoc) [reply go.mod:30]: Afaik this could be only `v1.2.3`, `v1.2.3` just got released https://github.com/openshift/imagebuilder/releases/tag/v1.2.3

**Review by flouthoc — APPROVED** [review]
LGTM

**Comment by giuseppe** [comment]
/hold

it seems like a regression introduced by 985eec53912a5afaabe3a4abdfa054231f03ce38.

**Comment by giuseppe** [comment]
I am still investigating.  Not sure why it doesn't happen in the CI for other PRs, but I can also reproduce the failure on origin/main:

```
$ git log -1
commit 1b8c014f3eee13d26d126b945959087629505396 (HEAD -> main, origin/main)
Merge: 494e0a9c 5aeb38bd
Author: OpenShift Merge Robot <openshift-merge-robot@users.noreply.github.com>
Date:   Tue Mar 29 14:19:46 2022 +0200

    Merge pull request #3871 from turrisxyz/naveensrinivasan/dependabot
    
    Updated dependabot to get updates for GitHub actions.
$ (cd copier; /usr/bin/go test -run TestGetPermissionErrorNoChroot)
--- FAIL: TestGetPermissionErrorNoChroot (0.03s)
    --- FAIL: TestGetPermissionErrorNoChroot/ignore=true (0.02s)
        copier_linux_test.go:149: 
            	Error Trace:	copier_linux_test.go:149
            	Error:      	Not equal: 
            	            	expected: 2
            	            	actual  : 4
            	Test:       	TestGetPermissionErrorNoChroot/ignore=true
            	Messages:   	expected two readable items, got 4
FAIL
exit status 1
FAIL	github.com/containers/buildah/copier	0.046s
```

**Pull request merged**
